### ISSUE TITLE
Mean reduction for non-autoregressive loss

### ISSUE THREAD

**Issue by shaochenze** [pytorch/fairseq#1668]
I notice that the loss function for non-autoregressive models uses a mean reduction: [code](https://github.com/pytorch/fairseq/blob/master/fairseq/criterions/nat_loss.py)
`nll_loss = mean_ds(losses)`
, where the loss for autoregressive models uses a sum reduction: [code](https://github.com/pytorch/fairseq/blob/master/fairseq/criterions/cross_entropy.py) 
`loss = F.nll_loss(
            lprobs,
            target,
            ignore_index=self.padding_idx,
            reduction='sum' if reduce else 'none',)`
Why is there such a difference? Do we need to set a higher learning rate if we use the mean reduction?

**Comment by myleott** [comment]
We use sum for autoregressive models but also pass a `sample_size` equal to the number of tokens, so the `Trainer` will normalize like this: `(sum of local gradient sums from each GPU) / (sum of sample_size across GPUs)`.

For nonautoregressive models we instead normalize the loss locally and then pass `sample_size=1` to normalize by the number of GPUs: https://github.com/pytorch/fairseq/blob/810e61ab3ffba169a47d7e72538e082d2467f2b4/fairseq/criterions/nat_loss.py#L123
Thus `Trainer` will normalize: `(sum of local gradient means from each GPU) / (number of GPUs)`

The two are nearly equivalent. The approach for autoregressive models is technically a bit more correct when the batch sizes (number of tokens) varies across GPUs, but often it's not possible to implement that efficiently so the approach taken by the nonautoregressive criterion works too.

**Comment by shaochenze** [comment]
Problem solved, thanks for the reply!
